storybookjs/storybook
# React: Add testing utilities

Issue: #10145

## What I did

Add a testing utility to @storybook/react to compose a story with decorators and parameters so that it can be easily reused in tests with tools like testing library.

This allow for codes in test files that would otherwise have quite some boilerplate to be as simple as:

1 - with `composeStories`, composing all stories
```jsx
import { render } from '@testing-library/react';
import { composeStories } from '@storybook/react';
import * as stories from './Button.stories';

// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level, meta level and global level.
const { Primary, Secondary } = composeStories(stories);

test('renders primary button with Hello World', () => {
  const { getByText } = render(<Primary>Hello world</Primary>); // You get props with autocompletion and you can override them, otherwise it will use the default values from the Story's args
  expect(getByText(/Hello world/i)).not.toBeNull();
});
```

2 - with `composeStory`, if you wish to do so for a single story
```jsx
import { render } from '@testing-library/react';
import { composeStory } from '@storybook/react';
import Meta, { Primary as PrimaryStory } from './Button.stories';

// Returns a component that already contain all decorators from story level, meta level and global level.
const Primary = composeStory(PrimaryStory, Meta);

test('renders primary button with Hello World', () => {
  const { getByText } = render(<Primary>Hello world</Primary>); // You get props with autocompletion and you can override them, otherwise it will use the default values from the Story's args
  expect(getByText(/Hello world/i)).not.toBeNull();
});
```

## Global config

If you have global decorators/parameters/etc and want them applied to your stories you need to first set it up. You can either do it in a setup file that should run once, before composeStory(ies) will be used, for instance:
```js
// setupFile.js <-- this will run before tests in jest.
import { setGlobalConfig } from '@storybook/react';
import * as globalStorybookConfig from './.storybook/preview';

setGlobalConfig(globalStorybookConfig);
```

Or you're free to use either API and pass a global config directly as a parameter, though I won't recommend because it requires more effort:

```jsx
import * as globalConfig from '../.storybook/preview';

const { Primary } = composeStories(stories, globalConfig);
const Primary = composeStory(PrimaryStory, Meta, globalConfig);
```

## How to test

1 - `yarn bootstrap` or `yarn build addons react --watch` if you prefer (addons just for the types, because react uses base types from addons)
2 - go to `examples/cra-ts-essentials`
3 - run `SKIP_PREFLIGHT_CHECK=true yarn test` (there's a compatibility problem with jest versions in the monorepo and the project, so just skip it for now and the tests should work)

- Is this testable with Jest or Chromatic screenshots?
Yes
- Does this need a new example in the kitchen sink apps?
Yes, added in `examples/cra-ts-essentials` project
- Does this need an update to the documentation?
Added, but I need feedback

## Further work

It's quite interesting to bring this to other major frameworks like Vue and Angular, but the way the tests are written is quite different (every framework with their specific peculiarities) and it would require further research and specific setup to allow them to take advantage of a testing utility like this one. I believe we should release this for React only first, get some feedback and then start working on other frameworks.
